core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
+{
+    use yansi::Paint as _;
 
-#[allow(non_upper_case_globals, non_snake_case)]
-impl Options {
-    /// All options disabled.
-    ///
-    /// Note that this is different than [`Options::default()`](#impl-Default),
-    /// which enables options.
-    pub const None: Options = Options(0);
+    let path = path.as_ref();
+    if !path.is_dir() {
+        let path = path.display();
+        error!("FileServer path '{}' is not a directory.", path.primary());
+        warn!("Aborting early to prevent inevitable handler error.");
+        panic!("invalid directory: refusing to continue");
+    }
+    let path = path.to_path_buf();
+    move |f| {
+        FileResponse::File(f.map_path(|p| path.join(p)))
+    }
+}
 
-    /// Respond to requests for a directory with the `index.html` file in that
-    /// directory, if it exists.
-    ///
-    /// When enabled, [`FileServer`] will respond to requests for a directory
-    /// `/foo` or `/foo/` with the file at `${root}/foo/index.html` if it
-    /// exists. When disabled, requests to directories will always forward.
-    ///
-    /// **Enabled by default.**
-    pub const Index: Options = Options(1 << 0);
+/// Prepends the provided path, to serve a single static file.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, file_root};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(file_root("static/index.html"))
+/// # }
+/// ```
+///
+/// # Panics
+///
+/// Panics if `path` does not exist.
+pub fn file_root(path: impl AsRef<Path>)
+    -> impl for<'p, 'h> Fn(File<'p, 'h>) -> FileResponse<'p, 'h> + Send + Sync + 'static
+{
+    use yansi::Paint as _;
 
-    /// Allow serving dotfiles.
-    ///
-    /// When enabled, [`FileServer`] will respond to requests for files or
-    /// directories beginning with `.`. When disabled, any dotfiles will be
-    /// treated as missing.
-    ///
-    /// **Disabled by default.**
-    pub const DotFiles: Options = Options(1 << 1);
+    let path = path.as_ref();
+    if !path.exists() {
+        let path = path.display();
+        error!("FileServer path '{}' is not a file.", path.primary());
+        warn!("Aborting early to prevent inevitable handler error.");
+        panic!("invalid file: refusing to continue");
+    }
+    let path = path.to_path_buf();
+    move |f| {
+        FileResponse::File(f.map_path(|p| path.join(p)))
+    }
+}
 
-    /// Normalizes directory requests by redirecting requests to directory paths
-    /// without a trailing slash to ones with a trailing slash.
-    ///
-    /// **Enabled by default.**
+/// Prepends the provided path, without checking to ensure the path exists during
+/// startup.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// # use rocket::fs::{FileServer, missing_root};
+/// # fn make_server() -> FileServer {
+/// FileServer::empty()
+///     .map_file(missing_root("/tmp/rocket"))
+/// # }
+/// ```
+pub fn missing_root(path: impl AsRef<Path>)
```

Comment:
I think we want to call this something different, say `file_root_permissive`, or something like that.